python-avd/pyavd/_eos_designs/structured_config/core_interfaces_and_l3_edge/utils.py

```diff
@@ -24,90 +27,72 @@ class UtilsMixin:
-        return get(self._hostvars, f"{self.data_model}.p2p_links_ip_pools", default=[])
-
-    @cached_property
-    def _p2p_links(self: AvdStructuredConfigCoreInterfacesAndL3Edge) -> list:
-        return get(self._hostvars, f"{self.data_model}.p2p_links", default=[])
-
     @cached_property
     def _p2p_links_sflow(self: AvdStructuredConfigCoreInterfacesAndL3Edge) -> bool | None:
-        return get(self._hostvars, f"fabric_sflow.{self.data_model}")
+        return self.inputs.fabric_sflow.core_interfaces if self.data_model == "core_interfaces" else self.inputs.fabric_sflow.l3_edge
 
     @cached_property
-    def _filtered_p2p_links(self: AvdStructuredConfigCoreInterfacesAndL3Edge) -> list:
+    def _filtered_p2p_links(self: AvdStructuredConfigCoreInterfacesAndL3Edge) -> list[tuple[T_P2pLinksItem, dict]]:
         """
         Returns a filtered list of p2p_links, which only contains links with our hostname.
 
         For each links any referenced profiles are applied and IP addresses are resolved
         from pools or subnets.
         """
-        if not (p2p_links := self._p2p_links):
+        if not (p2p_links := self.inputs_data.p2p_links):
             return []
 
         # Apply p2p_profiles if set. Silently ignoring missing profile.
-        if self._p2p_links_profiles:
-            p2p_links = [self._apply_p2p_links_profile(p2p_link) for p2p_link in p2p_links]
+        p2p_links: list[T_P2pLinksItem] = [self._apply_p2p_links_profile(p2p_link) for p2p_link in p2p_links]
 
         # Filter to only include p2p_links with our hostname under "nodes"
-        p2p_links = [p2p_link for p2p_link in p2p_links if self.shared_utils.hostname in p2p_link.get("nodes", [])]
+        p2p_links = [p2p_link for p2p_link in p2p_links if self.shared_utils.hostname in p2p_link.nodes]
         if not p2p_links:
             return []
 
         # Resolve IPs from subnet or p2p_pools.
         p2p_links = [self._resolve_p2p_ips(p2p_link) for p2p_link in p2p_links]
 
-        # Parse P2P data model and add simplified data
-        [p2p_link.update({"data": self._get_p2p_data(p2p_link)}) for p2p_link in p2p_links]
-
-        return p2p_links
+        # Parse P2P data model and create simplified data
+        return [(p2p_link, self._get_p2p_data(p2p_link)) for p2p_link in p2p_links]
 
-    def _apply_p2p_links_profile(self: AvdStructuredConfigCoreInterfacesAndL3Edge, target_dict: dict) -> dict:
-        """Apply a profile to a p2p_link."""
-        if "profile" not in target_dict:
+    def _apply_p2p_links_profile(self: AvdStructuredConfigCoreInterfacesAndL3Edge, p2p_link: T_P2pLinksItem) -> T_P2pLinksItem:
+        """Apply a profile to a p2p_link. Always returns a new instance. TODO: Raise if profile is missing."""
+        if not p2p_link.profile or p2p_link.profile not in self.inputs_data.p2p_links_profiles:
             # Nothing to do
-            return target_dict
+            return p2p_link._deepcopy()
```

Comment:
because we are still not behaving "nicely", so this ensures that we do not modify the original data. 
We can implement a "frozen" state on the new classes to catch all the places we modify (later).

Reply:
when it is frozen we can do it.